// yaml-language-server/out/server/test/bundlel10n.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const l10n = __importStar(require("@vscode/l10n"));
const assert_1 = __importDefault(require("assert"));
const path = __importStar(require("path"));
const node_1 = require("vscode-languageserver/node");
const schemaRequestHandler_1 = require("../src/languageservice/services/schemaRequestHandler");
const nodeTranslationSetup_1 = require("../src/nodeTranslationSetup");
const yamlServerInit_1 = require("../src/yamlServerInit");
const yamlSettings_1 = require("../src/yamlSettings");
const testHelper_1 = require("./utils/testHelper");
const testsTypes_1 = require("./utils/testsTypes");
describe('Bundle l10n Test', () => {
    let serverInit;
    before(() => {
        const yamlSettings = new yamlSettings_1.SettingsState();
        process.argv.push('--node-ipc');
        const connection = (0, node_1.createConnection)();
        const schemaRequestHandlerWrapper = (connection, uri) => {
            const testSchemaProvider = testHelper_1.TestCustomSchemaProvider.instance();
            const testSchema = testSchemaProvider.getContentForSchema(uri);
            if (testSchema) {
                return Promise.resolve(testSchema);
            }
            return (0, schemaRequestHandler_1.schemaRequestHandler)(connection, uri, yamlSettings.workspaceFolders, yamlSettings.workspaceRoot, yamlSettings.useVSCodeContentRequest, testHelper_1.testFileSystem, false);
        };
        const schemaRequestService = schemaRequestHandlerWrapper.bind(this, connection);
        const telemetry = new testsTypes_1.TestTelemetry(connection);
        serverInit = new yamlServerInit_1.YAMLServerInit(connection, yamlSettings, schemaRequestHandler_1.workspaceContext, schemaRequestService, telemetry, nodeTranslationSetup_1.setupl10nBundle);
    });
    after(async () => {
        await serverInit.setupl10nBundle({
            locale: 'en',
            processId: 0,
            rootUri: '',
            capabilities: undefined,
            initializationOptions: {
                l10nPath: path.join(__dirname, '../l10n'),
            },
        });
    });
    describe('l10n bundle test', function () {
        it('check french locale', async () => {
            await serverInit.setupl10nBundle({
                locale: 'fr',
                processId: 0,
                rootUri: '',
                capabilities: undefined,
                initializationOptions: {
                    l10nPath: path.join(__dirname, '../l10n'),
                },
            });
            assert_1.default.equal(l10n.t('Default value'), 'Valeur par défaut');
        });
        it('un configured locale should return in english', async () => {
            await serverInit.setupl10nBundle({
                locale: 'pt-br',
                processId: 0,
                rootUri: '',
                capabilities: undefined,
                initializationOptions: {
                    l10nPath: path.join(__dirname, '../l10n'),
                },
            });
            assert_1.default.equal(l10n.t('Default value'), 'Default value');
        });
    });
});
