// yaml-language-server/out/server/test/yamlCommands.test.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sinon = __importStar(require("sinon"));
const sinon_chai_1 = __importDefault(require("sinon-chai"));
const chai = __importStar(require("chai"));
const yamlCommands_1 = require("../src/languageservice/services/yamlCommands");
const commandExecutor_1 = require("../src/languageserver/commandExecutor");
const vscode_uri_1 = require("vscode-uri");
const expect = chai.expect;
chai.use(sinon_chai_1.default);
describe('Yaml Commands', () => {
    const JSON_SCHEMA_LOCAL = 'file://some/path/schema.json';
    const sandbox = sinon.createSandbox();
    let commandExecutorStub;
    beforeEach(() => {
        commandExecutorStub = sandbox.stub(commandExecutor_1.commandExecutor, 'registerCommand');
    });
    afterEach(() => {
        sandbox.restore();
    });
    it('should register handler for "JumpToSchema" command', () => {
        (0, yamlCommands_1.registerCommands)(commandExecutor_1.commandExecutor, {});
        expect(commandExecutorStub).to.have.been.calledWithMatch(sinon.match('jumpToSchema'), sinon.match.func);
    });
    it('JumpToSchema handler should call "showDocument"', async () => {
        const showDocumentStub = sandbox.stub();
        const getWorkspaceFoldersStub = sandbox.stub().returns(Promise.resolve([]));
        const connection = {
            window: {
                showDocument: showDocumentStub,
            },
            workspace: {
                getWorkspaceFolders: getWorkspaceFoldersStub,
            },
        };
        showDocumentStub.resolves(true);
        (0, yamlCommands_1.registerCommands)(commandExecutor_1.commandExecutor, connection);
        const arg = commandExecutorStub.args[0];
        await arg[1](JSON_SCHEMA_LOCAL);
        expect(showDocumentStub).to.have.been.calledWith({ uri: JSON_SCHEMA_LOCAL, external: false, takeFocus: true });
    });
    it('JumpToSchema handler should call "showDocument" with plain win path', async () => {
        const showDocumentStub = sandbox.stub();
        const getWorkspaceFoldersStub = sandbox.stub().returns(Promise.resolve([]));
        const connection = {
            window: {
                showDocument: showDocumentStub,
            },
            workspace: {
                getWorkspaceFolders: getWorkspaceFoldersStub,
            },
        };
        showDocumentStub.resolves(true);
        (0, yamlCommands_1.registerCommands)(commandExecutor_1.commandExecutor, connection);
        const arg = commandExecutorStub.args[0];
        await arg[1]('a:\\some\\path\\to\\schema.json');
        expect(showDocumentStub).to.have.been.calledWith({
            uri: vscode_uri_1.URI.file('a:\\some\\path\\to\\schema.json').toString(),
            external: false,
            takeFocus: true,
        });
    });
    it('JumpToSchema handler should call "showDocument" with plain POSIX path', async () => {
        const showDocumentStub = sandbox.stub();
        const getWorkspaceFoldersStub = sandbox.stub().returns(Promise.resolve([]));
        const connection = {
            window: {
                showDocument: showDocumentStub,
            },
            workspace: {
                getWorkspaceFolders: getWorkspaceFoldersStub,
            },
        };
        showDocumentStub.resolves(true);
        (0, yamlCommands_1.registerCommands)(commandExecutor_1.commandExecutor, connection);
        const arg = commandExecutorStub.args[0];
        await arg[1]('/some/path/to/schema.json');
        expect(showDocumentStub).to.have.been.calledWith({
            uri: vscode_uri_1.URI.file('/some/path/to/schema.json').toString(),
            external: false,
            takeFocus: true,
        });
    });
    it('JumpToSchema handler should call "showDocument" with custom web schema', async () => {
        const showDocumentStub = sandbox.stub();
        const getWorkspaceFoldersStub = sandbox.stub().returns(Promise.resolve([{ uri: 'vscode-test:///root/' }]));
        const connection = {
            window: {
                showDocument: showDocumentStub,
            },
            workspace: {
                getWorkspaceFolders: getWorkspaceFoldersStub,
            },
        };
        showDocumentStub.resolves(true);
        (0, yamlCommands_1.registerCommands)(commandExecutor_1.commandExecutor, connection);
        const arg = commandExecutorStub.args[0];
        await arg[1]('my-file.json');
        expect(showDocumentStub).to.have.been.calledWith({
            uri: vscode_uri_1.URI.parse('vscode-test:///root/my-file.json').toString(),
            external: false,
            takeFocus: true,
        });
    });
});
